docs/dcp2_system_design.rst

```diff
@@ -1223,49 +1247,117 @@ When a data release first enters the preparation phase, it is identical to the
+``delta`` staging area should be utilized.
+
+An ``update`` staging area contains a complete set of subgraphs, metadata
+entities, and descriptors. The ``entity_id`` of entities, and the
+``links_id`` of subgraphs, must be stable. Only genuinely new entities
+(or subgraphs) may be assigned a new ``entity_id`` (or ``links_id``). 
+
+When the partitioning of the graph into subgraphs changes, the staging area
+source should minimize ID churn: if the metadata graph is partitioned into
+100 subgraphs before an update, and 200 after, the source should write all
+200 subgraphs with a new ``links_id`` each. If instead the number of
+subgraphs remains the same but some or all of the subgraphs are subject to an
+update like the addition of an entity, the same ``links_id`` should be used
+for each updated subgraph. The ``links_id`` of unchanged subgraphs should
+remain the same, obviously. In between those two extremes (all new
+``links_id`` values or all the same), the source is free to pick any suitable
+algortithm for allocating new ``links_id`` values and reusing existing ones.
+
+Before importing an ``update`` staging area, the importer empties all entity
+and links tables in the *target* dataset. The *target* dataset is the TDR
+dataset into which the staging area is being imported. The importer does not
+delete any data files. It then proceeds importing the ``update`` staging area
+as if it were a ``complete`` staging area with one exception: If the importer
+encounters a descriptor referencing a data file that is absent from the
+staging area, the importer checks for an existing data file in the target
+dataset. If one exists and has the CRC32C or SHA-256 checksum specified in
+the descriptor, the importer continues to import the descriptor, referencing
+the pre-existing data file in the target dataset.
+
+Note that an ``update`` staging area may not alter the ``file_id``, ``size``,
+``crc32c`` and ``sha256`` properties, it may alter or add other properties
+such as ``file_name`` or ``describedBy``. If there is no existing data file
+in TDR or the checksums of an existing file don't match, the importer aborts
+the entire import with an error.
+
+Also note that there may still be data files in an ``update`` staging area and
+that those should be processed in the same manner as for ``complete`` staging
+areas.
+
+|nn| After importing an ``update`` staging area, the importer should check for
+orphans, and delete them. An orphan is a data file that was referenced by a
+row in a ``…_file`` table *before* the import, but not *after*. |ne|
+
+When the importer aborts an import due to, say, a checksum mismatch for a data
+file omitted from the imported staging area, it must restore the target
+dataset to the state before the aborted update.
+
+|nn| This could be implemented by rolling back prior changes, or by verifying
+the file checksums of all absent data files before making any modifications to
+the target dataset. |ne|
```

Comment:
@ncalvanese1, you may disagree with this.

Reply:
This one is a bit complicated, as the importer doesn't have an elegant means of rolling back changes, so if we had already cleared out the target dataset, our only option would be to reimport the "normal" staging area. Putting that aside, I could see value in moving the checksum verification up front rather than integrating it in the middle of the current workflow, so I'm actually okay with how this is written. What I would mostly want to avoid is a cycle of trial-and-error during a release. In other words, if we have to abort the update and effectively "restore" (retain) the target dataset, we shouldn't expect to be able to attempt a subsequent update for that dataset during that release.

Reply:
You could implement the file existence/checksum validation in https://github.com/DataBiosphere/hca-import-validation but it would mean that the validation tool can determine the existence and checksum of a file in a TDR dataset.

Reply:
>  our only option would be to reimport the "normal" staging area

Does this mean we can't delete them?

I wonder if it would be better to have different naming for updated staging areas so we won't have to touch the original staging areas we've originally used. Let's say `gs://bucket/dev/project-uuid-timestamp/` or `gs://bucket/dev/project-uuid-extra-id/` that we specify in the import form?

Reply:
> Does this mean we can't delete them?

I think this is why we would want to do the file existence validation before we go ahead and delete everything out of the dataset. Recovering the data otherwise becomes much more complicated.

Reply:
Staging areas were meant as temporary exchanges. If we've maneuvered ourselves into a place where staging areas can't be deleted, we have a problem. TDR is the long-term, authoritative data store. IOW

> our only option would be to reimport the "normal" staging area

is not a good option. How about repopulating the dataset from a previously cut snapshot?

Reply:
Also, I do seem to remember the mention of soft-deletes in conjunction with undoing imports of problematic staging areas. If the deletion is indeed soft, it should be reversible, correct? Or is the "emptying out of the dataset" as required by this spec change not done via soft-deletes?

Reply:
Apologies, I think there are still a few questions for me on here. I'll have to get to them early next week when I can carve out some time. Sorry about that!

Reply:
Apologies again for the delay here. 

RE: Staging area naming convention:  The staging area name gets used in a number of places in our code, and my preference would be to not change the naming convention for them.

RE: Soft-deletes: Yes, these deletes are "soft-deletes", so they are technically reversible. That being said, the process for doing so would likely be quiet manual and tedious, so I would prefer us to not rely on this if we can avoid it, but rather do validation upstream before we initiate the soft-deletes.